client/basic-authorship/src/basic_authorship.rs

```diff
@@ -210,7 +210,18 @@ impl<A, B, Block, C> ProposerInner<B, Block, C, A>
 		let mut is_first = true;
 		let mut skipped = 0;
 		let mut unqueue_invalid = Vec::new();
-		let pending_iterator = self.transaction_pool.ready();
+		let pending_iterator = match executor::block_on(future::select(
+			self.transaction_pool.ready_at(Some(self.parent_number)),
+			futures_timer::Delay::new(deadline - (self.now)()),
```

Comment:
That might not give us enough time to produce anything. I think it should be half of the remaining time to at least produce an empty block.

Reply:
I somewhat agree

But also there is an argument that we must just fail here instead of trying to recover from some invalid system state (transaction pool does not receive new block notifications or cannot handle them fast for some reason). Timeout is therefore here (for now) just for avoiding thread hanging forever, not for some logic after it.

Reply:
Updated to warning & continue